### PR TITLE
feat: JSON conversion improvements

### PR DESCRIPTION
**Pull request by eliotwrobson**
Improvements to json conversion that have been requested. Based on changes from #6744. Addresses the following:

- Resolves https://github.com/PrairieLearn/PrairieLearn/issues/1366
- Resolves https://github.com/PrairieLearn/PrairieLearn/issues/2099
- Resolves https://github.com/PrairieLearn/PrairieLearn/issues/6965

EDIT: Dependency on other PRs has been removed, should be good to review!

### CODE REVIEW

**Comment by coatless** [comment]
Eh, I'm a bit worried that we're now getting into the territory of a bunch of merge conflicts as #6744 changes have crept into this PR. :/

**Comment by eliotwrobson** [comment]
@coatless this is why this is in draft, I want to wait for other things to get merged before this is even reviewed seriously 😅

File: python/prairielearn.py
Comment (coatless) [inline python/prairielearn.py:162]: I don't think there is a way to incorporate a more specific error message here related to the problematic variable. e.g. 

```python
ValueError(f'Variable {variable_name} with type complex should have both with real and imaginary pair values inside of value.')
```

`variable_name` I think has to be passed in as there doesn't seem to be a good option to extract the information via `inspect`.

Comment (eliotwrobson) [reply python/prairielearn.py:162]: Yeah, this is kindof a persistent issue unfortunately. Debugging these json serialization issues is always a pain.

Regarding your other comment, since it seems like we might go in another direction with the dataframes, I'll remove those changes from this PR and flip it to open (should get merged sooner that way).

File: python/prairielearn.py
Comment (coatless) [inline python/prairielearn.py:780]: Why was the `to_json()` call removed here? Is `value` already in JSON format?

Comment (eliotwrobson) [reply python/prairielearn.py:780]: It's actually the opposite. This element was saving numpy floats natively because this to_json call wasn't doing anything. So adding the new encoding to this function actually broke some tests that were using custom grader code.

File: python/prairielearn.py
Comment (coatless) [inline python/prairielearn.py:78]: Let's add a quick variable name extractor: 

```suggestion
    # Retrieve the name of the `v` variable passed in by _inspecting_ the frame. 
    # Note, this approach does _not_ scale to multiple frame calls. Hence, we're opting not to functionalize it.
    variable_name = [k for k,x in inspect.currentframe().f_back.f_locals.items() if x is v][0]
    if np.isscalar(v) and np.iscomplexobj(v):
```

File: python/prairielearn.py
Comment (nwalters512) [inline python/prairielearn.py:81]: Are we doing something sensible if a number can't be represented as a 32-bit integer? IIRC Numpy ints are 64-bit but JSON can't represent a full 64-bit int.

File: python/prairielearn.py
Comment (nwalters512) [reply python/prairielearn.py:780]: Given this, it looks like the docs for this function are now out of date:

> On successful parsing, "data" will contain a 'submitted_answers' key that is the JSON encoded parsed answer.

File: python/prairielearn.py
Comment (eliotwrobson) [reply python/prairielearn.py:81]: Good catch! We can just always use a string to represent the integer value. I'll also add a test for this.

File: python/prairielearn.py
Comment (coatless) [reply python/prairielearn.py:136]: I thought about this a bit over break. I've surmised my thoughts here: 

https://github.com/PrairieLearn/PrairieLearn/issues/6883

Maybe we could have this as a _separate_ PR to follow once the updated JSON conversions are included.

File: python/prairielearn.py
Comment (eliotwrobson) [reply python/prairielearn.py:780]: Actually, looking at this function I think that the input string can be a complex number as well. I might try and see if there's a way to avoid changing this function. A little weird that sometimes it saves the submitted answer as a JSON dict with more detailed information and sometimes just an integer. Maybe the best way is to just update the custom grader tests?

Comment (eliotwrobson) [reply python/prairielearn.py]: @coatless after thinking about this some, do you think we should have the numpy encodings have an opt-out option (similar to the way the dataframe element is set up), to avoid issues with breaking peoples custom grader code relying on the way that numpy integers are stored? That way, the option can be applied to the `string_fraction_to_number` function, which will avoid breaking people's custom grader code.

File: python/prairielearn.py
Comment (coatless) [inline python/prairielearn.py:125]: Numbers are better than bools for new versions. 

```suggestion
    if np_encoding == 2 and type(v).__module__ == "numpy":
```

File: python/prairielearn.py
Comment (nwalters512) [inline python/prairielearn.py:99]: What's the migration path to the new encoding? Will we update all elements and example questions to use it? How do we communicate this change to courses? Is there a plan for deprecating and removing the legacy encoding entirely someday?

Is it strictly necessary to put this behind a flag? AFAICT the only case where this would cause issues is if someone was using `to_json` to store data but not using `from_json` to retrieve it, which feels like a question bug that we shouldn't feel the need to support.

If we decide we definitely need to flag this, the name `np_encoding` feels weird to me. I'd be inclined to make this either `np_encoding_version=1` or `legacy_np_encoding=True`.

Comment (eliotwrobson) [reply python/prairielearn.py:99]: Just to clarify, in a previous comment, I mistakenly said that numpy scalars were being auto-encoded as their corresponding Python types. No idea why I observed this particular behavior for a while, but PL just throws an error in this case as was noted in the original issue.

> What's the migration path to the new encoding? Will we update all elements and example questions to use it?

This encoding type is mainly for use by instructors writing custom questions (this seems to be the use case mentioned in the original issue #1366), so I don't think that there really needs to be a migration for core PL code. In particular, using this for the `pl-number-input` would change the way that some internal data is stored, which in turn would break custom grader code (including custom grader code used in internal test questions). So I think the best way to introduce this is by having it be opt-in for those that want it. 

> How do we communicate this change to courses?

Since this change is opt-in, I think we can communicate it through a news item of some type. There has been some discussion about a news item for the recent Python changes, and I'd be happy to whip up a combined one once we merge in some more of these instructor-facing changes.

> Is there a plan for deprecating and removing the legacy encoding entirely someday?

So the current legacy encoding is to do nothing (not sure why I thought that numpy types were being auto-encoded as their Python counterparts, testing again reveals that it just raises an error, as in the original issue), so I don't think there's much of a point to deprecating this. Alternatively, we could just make this particular encoding a different function altogether, but I think that could cause more confusion than it's worth.

> Is it strictly necessary to put this behind a flag? AFAICT the only case where this would cause issues is if someone was using `to_json` to store data but not using `from_json` to retrieve it, which feels like a question bug that we shouldn't feel the need to support.

As discussed above, there are cases where people (including core-PL code) run `to_json` and expect it to do nothing (don't know why these were the semantics that were decided upon, but that's how it works), and so the flag prevents breaking this in those cases.

> If we decide we definitely need to flag this, the name `np_encoding` feels weird to me. I'd be inclined to make this either `np_encoding_version=1` or `legacy_np_encoding=True`.

I think I'm leaning towards `np_encoding_version`, as @coatless has a preference against using a boolean flag.

Comment (eliotwrobson) [reply python/prairielearn.py:99]: Ok so I figured it out. Weirdly, the `float64` numpy type is json serializable (but get its type erased when decoding), and the other types are not. See [here](https://stackoverflow.com/questions/44459168/why-are-some-numpy-datatypes-json-serializable-and-others-not) for details. Either way, this means that the new conversion is technically backwards incompatible for some cases, so I think should be behind a flag (since some code relies on the above default behavior, including code in the number input element).

File: python/prairielearn_test.py
Comment (eliotwrobson) [inline python/prairielearn_test.py:139]: Had to change this because pytest was trying to pick this up as a test class before, whoops.

File: docs/question.md
Comment (nwalters512) [inline docs/question.md:243]: Minor wordsmithing:

```suggestion
The `to_json` function supports keyword-only options for different types of encodings (e.g. `to_json(var, df_encoding_version=2)`). These options have been added to allow for new encoding behavior while still retaining backwards compatibility with existing usage.

- `df_encoding_version` controls the encoding of Pandas DataFrames. Encoding a DataFrame `df` by setting `to_json(df, df_encoding_version=2)` allows for missing and date time values whereas `pl.to_json(df, df_encoding_version=1)` (default) does not. However, `df_encoding_version=1` has support for complex numbers, while `df_encoding_version=2` does not.

- `np_encoding_version` controls the encoding of Numpy values. When using `np_encoding_version=1`, then only `np.float64` and `np.complex128` can be serialized by `to_json`, and their types will be erased after deserialization (will become native Python `float` and `complex` respectively). It is recommended to set `np_encoding_version=2`, which supports serialization for all numpy scalars and does not result in type erasure on deserialization.
```

File: docs/question.md
Comment (nwalters512) [inline docs/question.md:224]:
```suggestion
All persistent data related to a question variant is stored under different entries in the `data` dictionary. This dictionary is stored in JSON format by PrairieLearn, and as a result, everything in `data` must be JSON serializable. Some types in Python are natively JSON serializable, such as strings, lists, and dicts, while others are not, such as complex numbers, numpy ndarrays, and pandas DataFrames.
```

**Review by nwalters512 — APPROVED** [review]

File: docs/question.md
Comment (coatless) [inline docs/question.md:226]:
```suggestion
To account for this, the `prairielearn` Python library from [`prairielearn.py`](https://github.com/PrairieLearn/PrairieLearn/blob/master/python/prairielearn.py), usually aliased and used as `pl`, provides the functions `to_json` and `from_json`, which can respectively serialize and deserialize various objects for storage as part of question data. Please refer to the docstrings on those functions for more information. Here is a simple example:
```

File: docs/question.md
Comment (coatless) [inline docs/question.md:241]:
```suggestion
The `pl.to_json` function supports keyword-only options for different types of encodings (e.g. `pl.to_json(var, df_encoding_version=2)`). These options have been added to allow for new encoding behavior while still retaining backwards compatibility with existing usage.
```

File: docs/question.md
Comment (coatless) [inline docs/question.md:243]:
```suggestion
- `df_encoding_version` controls the encoding of Pandas DataFrames. Encoding a DataFrame `df` by setting `pl.to_json(df, df_encoding_version=2)` allows for missing and date time values whereas `pl.to_json(df, df_encoding_version=1)` (default) does not. However, `df_encoding_version=1` has support for complex numbers, while `df_encoding_version=2` does not.
```

File: docs/question.md
Comment (coatless) [inline docs/question.md:245]:
```suggestion
- `np_encoding_version` controls the encoding of Numpy values. When using `np_encoding_version=1`, then only `np.float64` and `np.complex128` can be serialized by `pl.to_json`, and their types will be erased after deserialization (will become native Python `float` and `complex` respectively). It is recommended to set `np_encoding_version=2`, which supports serialization for all numpy scalars and does not result in type erasure on deserialization.
```

**Pull request merged**
